atlasdb-commons/src/main/java/com/palantir/common/concurrent/CoalescingSupplier.java

```diff
@@ -29,52 +29,61 @@
  * requested; requests will not receive results for computations that started prior to the request.
  */
 public class CoalescingSupplier<T> implements Supplier<T> {
-
     private final Supplier<T> delegate;
-    private volatile CompletableFuture<T> nextResult = new CompletableFuture<T>();
-    private final Lock fairLock = new ReentrantLock(true);
+    private volatile Round nextResult = new Round();
 
     public CoalescingSupplier(Supplier<T> delegate) {
         this.delegate = delegate;
     }
 
     @Override
     public T get() {
-        CompletableFuture<T> future = nextResult;
-
-        completeOrWaitForCompletion(future);
-
-        return getResult(future);
+        Round present = nextResult;
+        if (present.isFirstToArrive()) {
+            present.execute();
+            return present.getResult();
+        }
+        awaitDone(present.future);
+        Round next = present.next;
+        if (next.isFirstToArrive()) {
+            next.execute();
+        }
+        return next.getResult();
     }
 
-    private void completeOrWaitForCompletion(CompletableFuture<T> future) {
-        fairLock.lock();
-        try {
-            resetAndCompleteIfNotCompleted(future);
-        } finally {
-            fairLock.unlock();
+    private final class Round {
+        private final AtomicBoolean hasStarted = new AtomicBoolean(false);
+        private final CompletableFuture<T> future = new CompletableFuture<>();
+        private volatile Round next;
```

Comment:
it's unclear to me based on Java semantics whether this needs to be volatile or not. But I'm gonna leave it as such.

Reply:
I think not, since completing a future does happen before a join on that future. Not 100% confident in this one so agree with leaving it as such.